nasl/nasl_snmp.c

```diff
@@ -527,84 +628,179 @@ snmpv3_get (const char *peername, const char *username, const char *authpass,
 #endif /* HAVE_NETSNMP */
 
 static tree_cell *
-nasl_snmpv1v2c_get (lex_ctxt *lexic, int version)
+nasl_snmpv1v2c_get (lex_ctxt *lexic, int version, u_char action)
 {
-  const char *proto, *community, *oid_str;
-  char *result = NULL, peername[2048];
+  const char *proto;
+  char peername[2048];
   int port, ret;
+  snmpv1v2_request_t request;
+  snmp_result_t result;
+  char *oid_str;
+  static char *next_oid_str;
+
+  request = g_malloc0 (sizeof (struct snmpv1v2_request));
+  result = g_malloc0 (sizeof (struct snmp_result));
 
+  request->version = version;
+  request->action = action;
   port = get_int_var_by_name (lexic, "port", -1);
   proto = get_str_var_by_name (lexic, "protocol");
-  community = get_str_var_by_name (lexic, "community");
+  request->community = get_str_var_by_name (lexic, "community");
+
   oid_str = get_str_var_by_name (lexic, "oid");
-  if (!proto || !community || !oid_str)
-    return array_from_snmp_result (-2, "Missing function argument");
+  if (action == NASL_SNMP_GETNEXT && oid_str == NULL && next_oid_str != NULL)
+    request->oid_str = next_oid_str;
+  else
+    request->oid_str = oid_str;
+
+  if (!proto || !request->community || !request->oid_str)
+    return array_from_snmp_error (-2, "Missing function argument");
   if (port < 0 || port > 65535)
-    return array_from_snmp_result (-2, "Invalid port value");
+    return array_from_snmp_error (-2, "Invalid port value");
   if (!proto_is_valid (proto))
-    return array_from_snmp_result (-2, "Invalid protocol value");
+    return array_from_snmp_error (-2, "Invalid protocol value");
 
   g_snprintf (peername, sizeof (peername), "%s:%s:%d", proto,
               plug_get_host_ip_str (lexic->script_infos), port);
-  ret = snmpv1v2c_get (peername, community, oid_str, version, &result);
+  request->peername = peername;
+  ret = snmpv1v2c_get (request, result);
+
+  // Hack the OID string to adjust format. Replace 'iso.' with '.1.'
+  // This Allows to call getnext without an oid, since the last oid
+  // is stored.
+  if (result->oid_str != NULL && g_strstr_len (result->oid_str, 3, "iso"))
+    {
+      next_oid_str = result->oid_str + 2;
+      next_oid_str[0] = '1';
+      result->oid_str = g_strdup (next_oid_str);
+    }
+  else if (result->oid_str != NULL)
+    next_oid_str = result->oid_str;
+
+  /* Free request only, since members are pointers to the nasl lexic context
+     which will be free()'d later */
+  g_free (request);
+
   return array_from_snmp_result (ret, result);
```

Comment:
I guess the result struct also has to free its memory after the array is created

Reply:
No. Result struct must not be free()'d, since the return nasl tree_cell structs does not have a copy of the result members, but points to them. You can see this in the array_from_snmp_result().
Once the script finishes, the memory used by tree_cells structures is released.